### ISSUE TITLE
Future of community.docker

### ISSUE THREAD

**Issue by felixfontein** [ansible-collections/community.docker#364]
community.docker has been using the [Docker SDK for Python](https://github.com/docker/docker-py/) ("docker-py") for almost everything (directly for most content, and indirectly for the docker_compose module; only the docker connection plugin and the docker_stack modules use the CLI tool).

Unfortunately, it looks like he Docker SDK for Python is pretty much dead ([PR for adding unmaintained notice](https://github.com/docker/docker-py/pull/2989), [activity since October 2021](https://github.com/docker/docker-py/graphs/contributors?from=2021-10-12&to=2022-05-31&type=c)). Maybe a fork will come up, maybe they'll allow some community folks to take the project over under its current repository and PyPi package name, and maybe it will just die (as the Python docker-compose did).

Anyway, this is a moment to think about what we could and/or should do. Having a fork or the original project continued would obviously be the best solutions ;-) With a fork we might have to adjust the import statements (resp. allow to import both original docker-py and the fork), but (if they don't do something drastic) most things can just stay as they are. In case the project is dead and noone wants to maintain a fork, we do have a problem.

We have four different groups of contents:

1. Content that uses the Docker CLI program, or something else:
   * `docker` connection plugin
   * `nsenter` connection plugin (uses the `nsenter` CLI tool)
   * `docker_machine` inventory plugin (uses the `docker-machine` CLI tool)
   * `current_container_facts` module (inspects files)
   * `docker_stack*` modules

2. Docker content that uses the API part of the Docker SDK for Python (i.e. not the high-level functionality)
   * `docker_api` connection plugin
   * `docker_containers` inventory plugin
   * `docker_swarm` inventory plugin
   * `docker_config` module
   * `docker_container_exec` module
   * `docker_container_info` module
   * `docker_container` module
   * `docker_host_info` module
   * `docker_image_info` module
   * `docker_image_load` module
   * `docker_image` module
   * `docker_login` module
   * `docker_network_info` module
   * `docker_network` module
   * `docker_node_info` module
   * `docker_node` module
   * `docker_prune` module
   * `docker_secret` module
   * `docker_swarm_info` module
   * `docker_swarm` module
   * `docker_swarm_service_info` module
   * `docker_swarm_service` module
   * `docker_volume_info` module
   * `docker_volume` module

3. Docker content that uses the high-level functionality of the Docker SDK for Python
   * `docker_plugin` module

4. Docker content that uses the Python version of docker-compose
   * `docker_compose` module

**Comment by felixfontein** [comment]
I'm interested in at least keeping (most of) the Docker "plain" modules / plugins alive, probably by allowing them to work without the Docker SDK for Python. (I still don't use Swarm or Docker Compose, and I'm kind of glad especially about the latter.)

This will be a not be something easy, but I think it's worth it - also since it would allow to run the modules / plugins without installing anything.

**Comment by felixfontein** [comment]
@markuman that's what I mean with "work without the Docker SDK for Python" :)

**Comment by ssbarnea** [comment]
Directly related to https://github.com/docker/compose/issues/9510 - please comment there, maybe we can get few more months to perform a migration out of docker_compose if we can fix the too restrictive pinning of dependencies.

**Comment by ju2wheels** [comment]
If its going to be refactored completely, I have some bugs that I have yet to file on some of the commands where they are doing more than is necessary to complete the requested task based on the module arguments provided and in some cases that leads to failure where it really shouldnt.

I have to find my old tickets for the manual process on how to reproduce the issue, but basically one example I think was the docker_image task is unable to remove an image if you have overlayfs corruption issues (despite the docker-cli tools being able to do it perfectly fine) because it would collect info from the overlayfs that it didnt really need before actually using the API call to remove the image. I basically had to create a custom module that did nothing but call the API to remove the image to work around issues like this.

For us this corruption usually happens in our in field IOT devices while dowloading images that get interrupted due to bad network connectivity, so the corruption itself is a Docker side issue but the fact that we cant recover using the Ansible modules while we can recover with the docker-cli is a module implementation issue.

Food for thought.

**Comment by felixfontein** [comment]
@ju2wheels please file the bugs in new issues. This is not really related to this discussion.

**Comment by WojciechowskiPiotr** [comment]
I think we need to wait and see what the status of the Docker SDK will be. After all, it is maintained by the Docker team. Alternatives are to work directly with Docker API or use other SDK (probably not possible except for current SDK forks).

**Comment by ssbarnea** [comment]
That "need to wait"... until forever? ;) I doubt they are going to do anything TBH. Best bet would be to fork old compose and publish that fix under a different package name. After this work to migrate to use of cli for compose can be started.

**Comment by felixfontein** [comment]
I don't think anyone of us is interested in maintaining a fork of either Docker SDK for Python or Docker Compose.

**Comment by felixfontein** [comment]
One thing that could be interesting is vendoring the Docker SDK for Python into the collection, or at least parts of it. (That would also allow us to update it independently, and we could kill (some of) the code that handles different versions of it.) I've looked at how to implement some things yesterday and today, and especialy all the code that handles talking to the daemon (via Unix sockets, IP sockets, TLS, SSH, ...) will be painful to reimplement, while Docker SDK for Python already has pretty mature code for this. I created https://github.com/ansible-community/community-topics/issues/109 to discuss the licensing issues for this.

**Comment by InCogNiTo124** [comment]
What about creating commands by hand and invoking them through `os.system()`? There is not much user input to validate and sanitize so there is an option of constructing a command that would be ran.

Example:
```
community.docker.docker_compose:
  build: true
  nocache: true
  up: true
```

would become something along the lines of `os.system("docker compose build --no-cache && docker compose up -d")`

**Comment by felixfontein** [comment]
For docker_compose that's likely what needs to be done, though no idea how easy/hard that is to do, and no idea how check mode can be done this way.

For other modules, I would really, really avoid that.

**Comment by felixfontein** [comment]
> One thing that could be interesting is vendoring the Docker SDK for Python into the collection, or at least parts of it. (That would also allow us to update it independently, and we could kill (some of) the code that handles different versions of it.) I've looked at how to implement some things yesterday and today, and especialy all the code that handles talking to the daemon (via Unix sockets, IP sockets, TLS, SSH, ...) will be painful to reimplement, while Docker SDK for Python already has pretty mature code for this. I created [ansible-community/community-topics#109](https://github.com/ansible-community/community-topics/issues/109) to discuss the licensing issues for this.

Turns out the licensing discussion was already resolved ... one year ago. Damn, I completely forgot :) So including Apache 2.0 licensed content is totally fine as long as it is in module_utils (or complete modules), and you avoid importing GPL licensed code in it.

Simply vendoring the Docker SDK for Python won't work well since it is Python 3.6+ by now, while this collection stlil supports Python down to 2.6. (Even though you'll have to use a very old Docker SDK for Python for 2.6 support...) But we can extract the basic connection handling and util code from Docker SDK for Python and do not have to reimplement that. (And we can mix code from several versions of it, to get something that supports 2.6+.) I'll probably look into that soon...

**Comment by ssbarnea** [comment]
@felixfontein Thanks for helping with this, especially as every day passing is hurting molecule-docker users. IMHO, i would just drop support for old python and make a new release with vendored version as soon as possible.

If anyone wants to fix problems related to py2, they are welcome to spend some time doing so. Based on what I seen so far, there are not many around willing to help.

I will be quite happy to bring back the plugin to the list of supported plugins. Replacing the docker_compose module with command line calls can come later, when the pressure is smaller.

**Comment by aehlke** [comment]
I got bit really nastily by adopting this without knowing better, while still learning the Docker ecosystem. Would be great to put up big disclaimers about these being outdated/unmaintained (for now) in the documentation.

**Comment by felixfontein** [comment]
I finished a PR that vendors some code from Docker SDK for Python (adjusted to work with Python 2.7 and 3.5 as well), and adjusted all plain docker modules (except docker_container and docker_plugin) to use that code instead of relying on the Docker SDK for Python: #387 Reviews/testing/feedback are welcome. I'm also planning to adjust docker_container, but that is going to be more complex. I'll also work on the docker_api connection and the docker_containers inventory plugin. If someone else wants to work on docker_plugin or (some of) the Docker Swarm modules and inventory plugin, feel free to create follow-up PRs.

My general idea would be:
1. Release one more minor community.docker release with Python 2.6 deprecation (#388) and some licensing adjustments (#386 / #393 / whatever it will be).
2. Then bump `main` branch to the next major version, remove deprecations, drop support for Python 2.6, merge #387, and merge/work on follow-up PRs.
3. Once everyone is happy with the results, release 3.0.0 (maybe first one or more beta versions to gather more feedback).

Any opinions / comments / feedback?

**Comment by gotmax23** [comment]
+1 to deprecate Python 2.6 support. `ansible-core` 2.13 has already completely removed support for Python 2.6 targets. As the package maintainer for [ansible-collection-community-docker](https://src.fedoraproject.org/rpms/ansible-collection-community-docker) in Fedora, I would appreciate at least some support for the `stable-2` branch for a little while longer.

**Comment by felixfontein** [comment]
`stable-2` will pretty surely get bugfixes for some time (easily backportable things will get backported anyway, more complicated things... depends on whether someone is willing to backport it manually - I guess it depends a lot on what exactly it is). I think so far we backported every bugfix to `stable-1`, so I would assume this continues to happen if possible.

**Comment by felixfontein** [comment]
CC @Ajpantuso @dariko @DBendit @hannseman @imjoseangel @jgoldschrafe @jwitko @morph027 @porshkevich @ushuz @WojciechowskiPiotr @ximon18 as you have contributed new modules/plugins or stepped up as maintainers in the past, and might have so far missed this discussion. (If you're no longer interested in this collection, sorry for bothering you, simply ignore this discussion / unsubscribe from it :) )

**Comment by imjoseangel** [comment]
Thanks for sharing @felixfontein!

Didn't know about this move.

I understand from the #387 that we have moved to our own API client and daemon maintenance and we are changing all the modules to work with these.

To avoid overlapping on changes, where do you need help?

**Comment by felixfontein** [comment]
@imjoseangel thanks for your message! What would be helpful:
1. Comments on the general situation
2. Reviews and testing of (parts of) #387
3. If you're interested in the `docker_compose` module, it probably needs a complete rewrite to handle docker-compose >= 2.0.0. At the same time it needs to keep backwards compatibility with 1.x.y, so this is going to be a hard one. Maybe simply creating a new module for docker-compose >= 2.0.0 (say `docker_compose2`) would be better?
4. If someone wants to convert the `docker_plugin` module, that one is some more work to do
5. Then there are all the Docker Swarm modules:
   - `docker_config` module
   - `docker_node_info` module
   - `docker_node` module
   - `docker_secret` module
   - `docker_swarm_info` module
   - `docker_swarm` module
   - `docker_swarm_service_info` module
   - `docker_swarm_service` module

   I guess it would be great if these also would be converted (`docker_swarm_service` will probably be the hardest one though; the rest should be relatively easy). I don't use Swarm and don't plan on ever using it, so I won't work on these. If anyone is interested in converting them, feel free to work on them (either now based on my branch, or wait until my PR is merged and then work on them).

Fortunately all the other modules and plugins (especially the Docker Stack modules, that you @imjoseangel worked on so far) don't need any work, since they "only" work with the Docker CLI tool (or other CLI tools) that aren't affected by any of this.

**Comment by felixfontein** [comment]
When discussing this on IRC, we noticed we could also depreate some more API versions. #397 deprecates all API versions up to 1.24. API version 1.25 was supported by Docker 1.13, released in January 2017, and is the first version which comes with a machine-readable API description. Sticking to that one might make our live much easier in the future (no more looking at manually created API docs).

**Comment by gotmax23** [comment]
I looked into the package versions for Docker on different Linux distributions. Out of the currently supported distributions that I looked at, CentOS/RHEL 7 still has 1.13.x. The rest of the distributions are at least somewhat newer than that. Many users likely use the docker-ce repos instead, but I figured I'd at least point this out.

**Comment by felixfontein** [comment]
@gotmax23 thanks for checking!

I've just merged the Python 2.6 deprecation, and if nobody complains until this weekend, I'll merge the API < 1.25 deprecation (#397) and make a new release so the deprecations will be active. Afterwards I'll create `stable-2` and start preparing `main` for 3.0.0. The licensing adjustments will probably have to wait for another release.

If anyone is not happy with this timeline, please complain ASAP :)

**Comment by felixfontein** [comment]
Here are a couple of PRs that would be happy if they'd be reviewed:

- #363 and #399 bump `main`'s version to 3.0.0 and remove deprecations scheduled for 3.0.0.
- #398 is a subset of #387 which just has the vendored code; I'm planning to merge this first, and then create one PR for every plugin/module to be adjusted (and close #387). That should make reviewing easier.

**Comment by felixfontein** [comment]
The 2.7.0 release is in progress, `stable-2` exists, and from now on `main` is preparing for the 3.0.0 release.

**Comment by felixfontein** [comment]
I created the following PRs for the parts of #387:
- docker_container_exec module: #401
- docker_container_info module: #402
- docker_host_info module: #403
- docker_image_info module: #404
- docker_image_load module: #405
- docker_image module: #406
- docker_login module: #407
- docker_network_info module: #408
- docker_network module: #409
- docker_prune module: #410
- docker_volume_info module: #411
- docker_volume module: #412
- docker_containers inventory plugin: #413
- docker_api connection plugin: #414

**Comment by felixfontein** [comment]
My plan for the next steps is as follows:
1. Get the above PRs reviewed and merged. @briantist already reviewed all of them (thanks a lot for that!), I'll keep them open a few more days in case anyone else wants to take a look.
2. Once they are all merged, I'll create a 3.0.0-a1 alpha release. That will allow this to be tested more easily.
3. In the meantime I'm working on rewriting docker_container, and maybe someone wants to work on some of the other modules as well.
4. There should probably be at least one other pre-release (depending on how many people want to work on the remaining things, if any, maybe also multiple) so that the other rewritten modules (docker_container and maybe more) can also receive some testing, and potential problems found in the pre-releases before can be fixed and the fixes tested as well.
5. Once we had a pre-release with all fixes to all known problems out and nobody found a new one for some time, and nobody is working on rewriting another module anymore, we can do the final 3.0.0 release.

Does that sound good for everyone?

**Comment by GarrisonD** [comment]
Hi @felixfontein 

Are you aware of this https://github.com/moby/moby/issues/41711?
I have faced it today, see my comments at the bottom.

It looks like the Docker Engine HTTP API (`/build` endpoint at least) is as abandoned as `docker-py` is 😢 

Doesn't it feel like `community.docker` is migrating from one dead API to another?

**Comment by InCogNiTo124** [comment]
@GarrisonD I tried to play with your example and it's true. It builds correctly with `docker buildx build .` but it wont build in any way using the old build api

**Comment by GarrisonD** [comment]
@InCogNiTo124 the new API isn't that new actually.

I wanted to say that if this transition from `docker-py` takes place, wouldn't it make sense transitioning to gRPC that is used in Docker CLI and supported at the first place?

And yes, I understand that would be another level of contribution in terms of efforts.

**Comment by felixfontein** [comment]
> Doesn't it feel like `community.docker` is migrating from one dead API to another?

I don't think the HTTP based API is dead. But then, who knows. Docker changes their mind all the time anyway, so whatever I do there's a chance Docker will make it obsolete anyway, so :shrug: At least the HTTP API also works with podman...

**Comment by felixfontein** [comment]
So far I haven't seen any public documentation of such a "next gen" gRPC based API. From everything I've seen so far, it seems to be an internal API used by different components of Docker communicating with eacher other.

(That buildkit won't work with a simple HTTP based API such as the current Engine API is not really surprising when looking at the features it has, like more intelligently sending build context when needed. That's something that's impossible to implement without some RPC protocol, and picking gRPC for that instead of rolling your own sounds like a good idea. That doesn't mean that any other API will be replaced by a gRPC API though.)

**Comment by GarrisonD** [comment]
I didn't say that the HTTP API would be replaced with the gRPC one. I said that the HTTP API looks abandoned and lives in 2019 (or when BuildKit became publicly available?) like `docker-py`. Does migrating `build` command from `docker-py` to the HTTP API make sense then?

If you say that there are no public `gRPC` docs then the only solution is to run shell commands with Docker CLI. Also this sounds less complicated than using (implementing? 😨) gRPC...

I emphasize that it's only about `build` command. For other ones the HTTP API probably is fine.

Docker doesn't care about BuildKit support via the HTTP API. Ansible's Docker module is out-of-date now and this situation will get even worse as more interesting features will be added on top of BuildKit.

---

That wouldn't be a problem if I used `docker_image` directly. I would just replace it with a shell command 😆 but it's used in `molecule[docker]` and this way it becomes a problem 🙂

**Comment by felixfontein** [comment]
A 3.0.0-a1 prerelease has been released with the current state of the `main` branch.

**Comment by felixfontein** [comment]
I think my rewrite of docker_container is ready (#422 for anyone who is interested). Is anyone interested in rewriting docker_plugin, or any of the Swarm modules? Or wants to work on docker_compose?

**Comment by ssbarnea** [comment]
I found it, I need to add `--pre`, and that is from inside ansible-compat. The bug is in ansible-galaxy cli which is not able to infer the fact that the range requires a pre-release, like pip does.

That prevents me from using a pre-release inside molecule-docker...

**Comment by felixfontein** [comment]
I've created another WIP PR #429 to adjust docker_plugin. Then all non-Swarm plugins and modules (that use the Docker SDK for Python) have been converted.

**Comment by felixfontein** [comment]
docker_container and docker_plugin have been converted, and docker_container has a few new features that weren't possible with Docker SDK for Python. I've created a new prerelease 3.0.0-a2 with all latest changes. Please test it thoroughly so we can weed out as many potential bugs before the 3.0.0 release as possible :)

**Comment by ssbarnea** [comment]
I plan to fix molecule-docker plugin over this weekend, so that should give a good testing ground for the new version.

**Comment by felixfontein** [comment]
I've ceated a 3.0.0-rc1 release with one bugfix (the only bug reported so far). If there are no reports in, say, ~two weeks, I'll create the final 3.0.0 release.

**Comment by felixfontein** [comment]
Looks like Docker SDK for Python is back to being actively maintained, judging from the list of commits to `main` in the last three days (one full page): https://github.com/docker/docker-py/commits/main

**Comment by felixfontein** [comment]
I created #435 to keep track of changes to Docker SDK for Python that might need to be cherry-picked to our vendored code.

**Comment by felixfontein** [comment]
I created a 3.0.0-rc2 release chandidate including the changes from Docker SDK for Python. If there are no further bug reports, 3.0.0 will be released in 1-2 weeks.

**Comment by gotmax23** [comment]
https://github.com/gabrieldemarmiesse/python-on-whales (python-on-whales on PyPI) provides a Python interface to compose v2 (the `docker compose` cli plugin), amongst other things. This is a community project, but it was featured on the Docker blog last year https://www.docker.com/blog/guest-post-calling-the-docker-cli-from-python-with-python-on-whales/. It shells out to the docker binary, but that's really the only option with the current `docker compose` interface.

**Comment by felixfontein** [comment]
I've seen that library before. There are two problems with it:
1. It's yet another library that users would have to install to use the module;
2. It requires Python 3.7+ (https://github.com/gabrieldemarmiesse/python-on-whales/blob/master/setup.py#L22).

So if the docker_compose module would start using it to talk to docker-compose 2.0+, it would still need to be able to support docker-compose < 2.0.0 without this library installed, and on Python 2.7, 3.5, and 3.6.

Alternatively, maybe it might be interesting to create a new module `docker_compose2` which only talks to docker-compose 2.0+. Such a new module could have different requirements than the old module, and it wouldn't need to support both docker-compose < 2.0 and 2.0+ at the same time.

On the other hand, I'm not sure how useful the python-on-whales library actually is for this task, since composing a command line and calling an external CLI program isn't something that hard to do manually / with `module.run_command()`. But I guess that's something that someone has to decide who actually wants to work on docker-compose support. I'm not using it at all, so I probably won't work on it.

**Comment by ssbarnea** [comment]
Give what happened during the last years, I would say that probably the best bet is to just call "docker compose" cli. I am not concerned so much about backwards compatibility because that would be insanely hard to achieve and maintain, likely impossible in some cases too. If anyone need to use old dockers, they should stick with old libraries/modules/collections. Extra libraries in-between would only add another thing to maintain.

Usually I am in favour of using APIs, but when it comes to docker, it seems that only stable api is "cli".

I find it funny that it took docker team years to put some fixed to old docker python library, and that happens only after we decided to go this way.

**Comment by maica1** [comment]
It seens like the last days had a lot of news, first we have the [python-on-whales ](https://www.docker.com/blog/guest-post-calling-the-docker-cli-from-python-with-python-on-whales/)project that uses the CLI tool under the hood.
And some activity on the docker-py repo, with new releases and the closing of the  [Add unmaintained notice](https://github.com/docker/docker-py/pull/2989)  issue
So maybe we can see some progress on the community.

**Comment by felixfontein** [comment]
I'm currently planning to do the 3.0.0 release this weekend, assuming nobody wants to veto that :)

**Comment by felixfontein** [comment]
I've just released 3.0.0 :tada: I'm closing this issue for now. If someone wants to discuss some aspects we haven't discussed / completed yet, please create a new issue for this. Thanks everyone!
